File: randomizer-compiler/src/main/java/com/kostasdrakonakis/randomizer/compiler/RandomProperty.java
package com.kostasdrakonakis.randomizer.compiler;

import javax.lang.model.element.Element;
import javax.lang.model.element.Name;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

abstract class RandomProperty {
    Element element;
    Name qualifiedClassName;
    Name simpleClassName;
    Name elementName;
    TypeMirror elementType;

    public RandomProperty(Element element) {
        this.element = element;
        elementName = element.getSimpleName();
        simpleClassName = element.getEnclosingElement().getSimpleName();
        TypeElement enclosingElement = ((TypeElement) element.getEnclosingElement());
        qualifiedClassName = enclosingElement.getQualifiedName();
        elementType = element.asType();
    }

    Name getQualifiedClassName() {
        return qualifiedClassName;
    }

    Name getSimpleClassName() {
        return simpleClassName;
    }

    Name getElementName() {
        return elementName;
    }

    TypeMirror getElementType() {
        return elementType;
    }

    Element getElement() {
        return element;
    }

    abstract boolean isTypeValid(Elements elements, Types types);

    abstract String getRandomValue();
}

File: randomizer-compiler/src/main/java/com/kostasdrakonakis/randomizer/compiler/RandomProperty.kt
package com.kostasdrakonakis.randomizer.compiler

import javax.lang.model.element.Element
import javax.lang.model.element.Name
import javax.lang.model.element.TypeElement
import javax.lang.model.type.TypeMirror
import javax.lang.model.util.Elements
import javax.lang.model.util.Types

internal abstract class RandomProperty(var element: Element) {
    val qualifiedClassName: Name
    val simpleClassName: Name = element.enclosingElement.simpleName
    val elementName: Name = element.simpleName
    private val elementType: TypeMirror

    abstract fun isTypeValid(elements: Elements, types: Types): Boolean
    abstract val randomValue: String

    init {
        val enclosingElement = element.enclosingElement as TypeElement
        qualifiedClassName = enclosingElement.qualifiedName
        elementType = element.asType()
    }
}